scipy/special/stirling2.h

```diff
@@ -0,0 +1,114 @@
+#ifndef STIRLING_H
+#define STIRLING_H
+
+/* c implementation of Stirling numbers of the second kind */
+
+#include <complex.h>
+#include <math.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+/*
+ *
+ *     Stirling numbers of the second kind
+ *
+ *
+ *
+ * SYNOPSIS: Stirling numbers of the second kind count the
+ *  number of ways to make a partition of n distinct elements
+ *  into k non-empty subsets.
+ *
+ * DESCRIPTION: n is the number of distinct elements of the set
+ *  to be partitioned and k is the number of non-empty subsets.
+ *  The values for n < 0 or k < 0 are interpreted as 0. If you
+ *
+ * ACCURACY: The method returns a double type.
+ *
+ * NOTE: this file is *NOT* part of the originaldistribution and was
+ *  added by Lucas Roberts
+ */
+
+// Dynamic programming approach
+
+double stirling2_dp(int n, int k){
+    if ((n == 0 && k == 0) || (n==1 && k==1)) {
+        return 1.;
+    }
+    if (k <= 0 || k > n || n < 0){
+        return 0.;
+    }
+    int arraySize = k <= n - k + 1 ? k : n - k + 1;
+    double *curr = (double *) malloc(arraySize * sizeof(double));
+    for (int i = 0; i < arraySize; i++){
+        curr[i] = 1.;
+    }
+    if (k <= n - k + 1) {
+        for (int i = 1; i < n - k + 1; i++){
+            for (int j = 1; j < k; j++){
+                curr[j] = (j + 1) * curr[j] + curr[j - 1];
+                // supported in c99: https://devdocs.io/c/numeric/math/isinf
+                if (isinf(curr[j])){
+                    free(curr);
+                    return INFINITY; // numeric overflow
+                }
+            }
+        }
+    } else {
+        for (int i = 1; i < k; i++){
+            for (int j = 1; j < n - k + 1; j++){
+                curr[j] = (i + 1) * curr[j - 1] + curr[j];
+                // supported in c99: https://devdocs.io/c/numeric/math/isinf
+                if (isinf(curr[j])){
+                    free(curr);
+                    return INFINITY; // numeric overflow
+                }
+            }
+        }
+    }
+    double output = curr[arraySize - 1];
+    free(curr);
+    return output;
+}
+
+
+// second order Temme approximation
+double stirling2_temme(int n, int k){
+  if ((n == k && n >= 0) || (n > 0 && k==1)){
+      return 1.;
+  }
+  if (k <= 0 || k > n || n < 0){
+      return 0.;
+  }
+  double mu = k / (double)n;
+  double delta = 1. / mu * pow(exp(-1), (-1/mu));
+  // note: lambert returns complex value, we only want the real part
+  double x0 = 1. / mu;
+  // x0 += lambertw(-delta);
+  double t0 = (n - k) / (double)k;
+  double F = sqrt(t0/((1 + t0)*(x0 - t0)));
+  double A = -n * log(x0) + k * log(pow(exp(-1), x0) - 1) - k * t0 + (n - k) * log(t0);
+  double F1 = -2*x0*x0*x0 + 2*t0*t0*t0*t0*t0 + 4*t0*t0*t0 + 4*t0*t0*t0*t0;
+  F1 += 3*x0*x0 * t0 - 6*x0*t0*t0*t0*t0;
+  F1 += -5*x0*x0 * t0*t0 + 2*x0*x0*x0*x0*t0 + x0*x0*x0*t0 - 6*x0*x0*x0 + t0*t0;
+  F1 += 8*x0*x0 * t0*t0*t0 / (24*F*(1 + t0) * (1 + t0)*(x0 - t0)*(x0 - t0)*(x0 - t0)*(x0-t0));
```

Comment:
We could some operations by doing some [Horner's method](https://en.wikipedia.org/wiki/Horner%27s_method) style manipulations here.

Reply:
push `7ca1aa6` has these changes

Reply:
Nice!